### PR TITLE
stats: Add iterate methods for tags.

### PR DESCRIPTION
**Pull request by jmarantz**
Description: With #6161, the tags are no longer held in memory as a vector<Tags>, but that structure has to be created when tags() is called. If a caller's goal is to find a tag by name, it's more efficient to capture the tag-name in a StatName in a context, and then iterate over the tags as StatNames looking for it. This provides two variants on an iterate() method to make this easier: the very efficient iterateTagStatNames(), and a slightly more compute-intensive version that iterates over const Tag& objects.
Risk Level: low
Testing: //test/common/stats/... -- note that tags() is already called in various tests, and that depends on all the new method bodies.
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by jmarantz** [comment]
@eziskind this is intended to provide a higher-performing way for stats sinks to look for a particular stat, rather than building a temp array of string-pairs, per f2f discussion.

File: source/common/stats/metric_impl.cc [hunk 1]
```diff
@@ -46,33 +46,53 @@ StatName MetricImpl::tagExtractedStatName() const {
...
+  });
+}
+
+std::vector<Tag> MetricImpl::tags() const {
+  std::vector<Tag> tags;
+
+  // StatNameList maintains a linear ordered collection of StatNames, and we
+  // are mapping that into a tag-extracted name (the first element), followed
+  // by alternating TagName and TagValue. So we use a little state machine
+  // as we iterate through the stat_names_.
```
Comment (eziskind) [inline source/common/stats/metric_impl.cc]: This comment looks out of place (it's copy-pasted from the iterateTagStatNames method).

**Comment by eziskind** [comment]
> Add unit tests - otherwise LGTM, Thanks!

Ah, I see your comment about how the new methods are already tested with unit tests that use "tags". So I guess you don't need to add more

File: source/common/stats/metric_impl.cc [hunk 2]
```diff
@@ -46,33 +46,53 @@ StatName MetricImpl::tagExtractedStatName() const {
   return stat_name;
 }
 
-std::vector<Tag> MetricImpl::tags() const {
-  std::vector<Tag> tags;
+void MetricImpl::iterateTagStatNames(const TagStatNameIterFn& fn) const {
   enum { TagExtractedName, TagName, TagValue } state = TagExtractedName;
-  Tag tag;
-  const SymbolTable& symbol_table = symbolTable();
+  StatName key;
```
Comment (zuercher) [inline source/common/stats/metric_impl.cc]: nit: Maybe call this tag_name?

Comment (jmarantz) [reply source/common/stats/metric_impl.cc]: yup, done.

**Comment by jmarantz** [comment]
Thanks for the re-approve @zuercher  :)

**Pull request merged**